File: serialization.py
"""
input and output some tricky matrix data in a cross-language format
"""

from scipy.sparse import coo_matrix, dok_matrix, csc_matrix
import tables
import numpy as np

def write_sparse_hdf(handle, group, data, colnames=None, filt=None):
    """
    sparse CSC (compressed sparse colums) matrices via hdf5
    32 bit for now, ok?
    """
    data_atom_type = tables.Float32Atom()
    if np.issubdtype(data.dtype, int):
        data_atom_type = tables.Int32Atom()
    handle.create_carray(group,'v_indices',
        atom=tables.Int32Atom(), shape=data.indices.shape,
        title="indices",
        filters=filt)[:] = data.indices
    handle.create_carray(group,'v_indptr',
        atom=tables.Int32Atom(), shape=data.indptr.shape,
        title="index ptr",
        filters=filt)[:] = data.indptr
    handle.create_carray(group,'v_data',
        atom=data_atom_type, shape=data.data.shape,
        title="data",
        filters=filt)[:] = data.data
    handle.create_carray(group,'v_datadims',
        atom=tables.Int32Atom(), shape=(2,),
        title="data dims",
        filters=filt)[:] = data.shape
    if colnames:
        handle.create_carray(group,'v_col_names',
            atom=tables.StringAtom(
                max([len(n) for n in colnames])
            ), shape=(len(colnames),),
            title="col names",
            filters=filt)[:] = colnames

def read_sparse_hdf(handle, group):
    """
    sparse CSC (compressed sparse colums) matrices via hdf5
    """
    shape = group.get_node(group,'v_datadims')
    return csc_matrix(
        (
            group.get_node(group,'v_data'),
            (
                group.get_node(group,'v_indices'),
                group.get_node(group,'v_indptr')
            )
        ),
        shape=shape)
    # if colnames:
    #     handle.create_carray(group,'v_col_names',
    #         atom=tables.StringAtom(
    #             max([len(n) for n in colnames])
    #         ), shape=(len(colnames),),
    #         title="col names",
    #         filters=filt)[:] = colnames

def read_cv_glmnet(handle, group):
    cv_lo = handle.get_node(group, 'v_cvlo').read()
    cv_up = handle.get_node(group, 'v_cvup').read()
    cv_sd = handle.get_node(group, 'v_cvsd').read()
    cv_m = handle.get_node(group, 'v_cvm').read()
    nzero = handle.get_node(group, 'v_nzero').read()
    nulldev = handle.get_node(group, 'v_nulldev').read()
    lambd = handle.get_node(group, 'v_lambda').read()
    coef_path = handle.get_node(group, 'v_coef').read()

    return d(
        cv_lo = cv_lo,
        cv_up = cv_up,
        cv_sd = cv_sd,
        cv_m = cv_m,
        nzero = nzero,
        nulldev = nulldev,
        lambd = lambd,
        coef_path = coef_path,
    )

File: serialization.R
require(rhdf5)

load.sparse.hdf = function (filename, path) {
  idx = as.vector(h5read(filename, paste(path, "v_indices", sep="/")))
  idxptr = as.vector(h5read(filename, paste(path, "v_indptr", sep="/")))
  vals = as.vector(h5read(filename, paste(path, "v_data", sep="/")))
  dims = as.vector(h5read(filename, paste(path, "v_datadims", sep="/")))
  col.names = h5read(filename, paste(path, "v_col_names", sep="/"))
  
  data = sparseMatrix(
    i=idx,
    p=idxptr,
    x=vals,
    dims=dims,
    index1=F
  )
  colnames(data)=col.names
  return(data)
}

save.glmnet.hdf = function (filename, path, cv.fit) {
  h5createGroup(filename, path)
  h5write.default(cv.fit$nzero, filename, paste(path, "v_nzero", sep="/"))
  h5write(cv.fit$cvlo, filename, paste(path, "v_cvlo", sep="/"))
  h5write(cv.fit$cvup, filename, paste(path, "v_cvup", sep="/"))
  h5write(cv.fit$cvsd, filename, paste(path, "v_cvsd", sep="/"))
  h5write(cv.fit$lambda, filename, paste(path, "v_lambda", sep="/"))
  h5write(cv.fit$cvm, filename, paste(path, "v_cvm", sep="/"))
  if (!is.null(cv.fit$glmnet.cv.fit$dev.ratio)) {h5write(cv.fit$glmnet.cv.fit$dev.ratio, filename, paste(path, "v_nulldev", sep="/"))}
  coef.mat = matrix(data=0,nrow=length(cv.fit$lambda), ncol=length(coef(cv.fit)))
  colnames(coef.mat) =rownames(coef(cv.fit))
  for (i in 1:length(cv.fit$lambda)) { l=cv.fit$lambda[i]; coef.mat[i,]=t(as.matrix(coef(cv.fit,s=l)))}
  #beware! includes new intercept term, and there is no colnames attribute to make this clear
  # > h5write(coef.mat, filename, "v_coef",  write.attributes = T)
  ## Warning message:
  ##In h5writeAttribute.default(Attr[[i]], h5obj, name = names(Attr)[i]) :
  ##  No function found to write attribute of class 'list'. Attribute 'dimnames' is not written to hdf5-file.
  h5write(coef.mat, filename,  paste(path, "v_coef", sep="/"))
}
